oauth2_provider/models.py

```diff
@@ -202,6 +215,13 @@ def clean(self):
                     grant_type=self.authorization_grant_type
                 )
             )
+        allowed_origins = self.allowed_origins.strip().split()
+        if allowed_origins:
+            # oauthlib allows only https scheme for CORS
+            validator = URIValidator({"https"})
```

Comment:
The URIValidaor({"https"}) isn't a good fit here.  We need to allow testing on localhost for developers. IIRC origins should not have a path or fragment component but should only be <scheme>://<hostname>:<port>.  We probably need to implement an AllowedOriginValidator that extends the RedirectURIValidator with and allow_paths property and check. 

I'm a little confused as to why the `RedirectUriValidator` isn't instantiated as `RedirectURIValidator(WildcardSet(), allowed_schemes=allowed_schemes)` and `allowed_schemes` are being checked explicitly when scheme validation is part of `URLValidator`. We don't need to fix it as a part of this PR but it would be nice to put a 
```python 
# TODO:  Document why we're are not passing  allowed_schemes to RedirectURI Validation and explicitly checking. 
# If there isn't a reason we should rewrite this to use the allowed_schemes validation built into the parent class.
```

Reply:
@dopry good point for localhost testing. Initially I was supposing that oauthlib allows only `https` origins.  (https://github.com/oauthlib/oauthlib/blob/master/oauthlib/oauth2/rfc6749/grant_types/base.py#L256) :
```
if not is_secure_transport(origin):
    log.debug('Origin "%s" is not HTTPS, CORS not allowed.', origin)
    return {}
```
But then discovered that setting `OAUTHLIB_INSECURE_TRANSPORT` environment variable will allow any schema. (https://github.com/oauthlib/oauthlib/blob/master/oauthlib/oauth2/rfc6749/utils.py#L81)
This should be mentioned in documentation, otherwise user experience will be confusing, because non-https origins will be ignored by default.

Reply:
@akanstantsinau thanks for finding that.  We should definitely mention it in the docs. Right now I'm wondering what is a good approach for allowed schemes. Application.get_allowed_schemes seems to be universal, but the setting ALLOWED_REDIRECT_URI_SCHEMES seems to be redirect specific.  It also enables ['http'] by default, which I feel is something a consumer should need to have to manually setup since it is insecure...   Part of me thinks we should add an ALLOWED_SCHEMES setting that is universal and use it for origins and set OAUTHLIB_INSECURE_TRANSPORT based on ALLOWED_SCHEMES...  In the future we can look at deprecating ALLOWED_REDIRECT_URI_SCHEMES in favor of an ALLOWED_SCHEMES....  We allow other schemes mainly for native app support, see: https://www.oauth.com/oauth2-servers/redirect-uris/redirect-uris-native-apps/

Reply:
Unless we're promoting really insecure defaults, IMO:

1. only allow `https`; but _also_ `http://localhost` / `http://127.0.0.1` / `http://::1` (plus dynamic tcp ports). MS Azure AD, for example, operates this way.
3. or someone very manually opts into `http` for non-localhost

I don't think people should have to allow all `http` to enable development/test setups using `localhost`. I can't see why CORS should be any different to redirect URIs in this regard?

Reply:
As a sysadmin I want to ensure that no one can use an http:// origin, localhost or otherwise, on my production server.   We only allows testing against our production Auth server from SSL secured UAT environments.

Reply:
So do I in production, but I don't have external developer apps either :-) 

So feels like 3 options? 
* `https`: nothing but
* `http`: yolo security
* https with localhost-http: `auto`, `https+local`, `balance`, `default`, `mixed`... ? naming is hard.

Anyway, that change could land separately, since it presumably would apply to both origins+redirect URIs via ALLOWED_SCHEMES

Reply:
Maybe we can open a feature request to discuss those settings. Personally I tend to prefer setting explicit values like `ALLOWED_SCHEMES=http,https,myapp` , `ALLOWED_DOMAINS=/regex/` over magic values like `auto`, `https+local`, `balance`, `default`, `mixed`. I find presets creating code that we need to maintain that isn't directly related to the business of authenticating, and they hide what's going on under the hood from consumers of the lib.

Reply:
@akanstantsinau sorry about going on a tangent. We still need to resolve the Origin Validation issue. Are you up to adding a OriginURIValidator and oauth2_settings.ALLOWED_SCHEMES that will ensure the original is just <scheme>://<host>:<port>?